src/saveload/town_sl.cpp

```diff
@@ -113,7 +113,89 @@ void UpdateHousesAndTowns()
 	RebuildTownCaches();
 }
 
-/** Save and load of towns. */
+class SlTownSupplied : public DefaultSaveLoadHandler<SlTownSupplied> {
+public:
+	inline const static SaveLoad description[] = {
+		SLE_CONDVAR(TransportedCargoStat<uint32>, old_max, SLE_UINT32, SLV_165, SL_MAX_VERSION),
+		SLE_CONDVAR(TransportedCargoStat<uint32>, new_max, SLE_UINT32, SLV_165, SL_MAX_VERSION),
+		SLE_CONDVAR(TransportedCargoStat<uint32>, old_act, SLE_UINT32, SLV_165, SL_MAX_VERSION),
+		SLE_CONDVAR(TransportedCargoStat<uint32>, new_act, SLE_UINT32, SLV_165, SL_MAX_VERSION),
+	};
+
+	void Save(void *object) override
+	{
+		Town *t = static_cast<Town *>(object);
+
+		for (CargoID i = 0; i < NUM_CARGO; i++) {
+			SlObject(&t->supplied[i], this->GetDescription());
+		}
+	}
+
+	void Load(void *object) override
+	{
+		Town *t = static_cast<Town *>(object);
+
+		uint num_cargo = IsSavegameVersionBefore(SLV_EXTEND_CARGOTYPES) ? 32 : NUM_CARGO;
+		for (CargoID i = 0; i < num_cargo; i++) {
+			SlObject(&t->supplied[i], this->GetDescription());
+		}
+	}
+};
+
+class SlTownReceived : public DefaultSaveLoadHandler<SlTownReceived> {
+public:
+	inline const static SaveLoad description[] = {
+		SLE_CONDVAR(TransportedCargoStat<uint16>, old_max, SLE_UINT16, SLV_165, SL_MAX_VERSION),
+		SLE_CONDVAR(TransportedCargoStat<uint16>, new_max, SLE_UINT16, SLV_165, SL_MAX_VERSION),
+		SLE_CONDVAR(TransportedCargoStat<uint16>, old_act, SLE_UINT16, SLV_165, SL_MAX_VERSION),
+		SLE_CONDVAR(TransportedCargoStat<uint16>, new_act, SLE_UINT16, SLV_165, SL_MAX_VERSION),
+	};
+
+	void Save(void *object) override
+	{
+		Town *t = static_cast<Town *>(object);
+
+		for (int i = TE_BEGIN; i < TE_END; i++) {
+			SlObject(&t->received[i], this->GetDescription());
+		}
+	}
+
+	void Load(void *object) override
+	{
+		Town *t = static_cast<Town *>(object);
+
+		for (int i = TE_BEGIN; i < TE_END; i++) {
+			SlObject(&t->received[i], this->GetDescription());
+		}
+	}
+};
+
+class SlTownAcceptanceMatrix : public SaveLoadHandler {
+public:
+	SaveLoadTable GetDescription()
+	{
+		/* Here due to private member vars. */
```

Comment:
Is this still needed? I can't see `GetTileMatrixDesc` being made `friend` anywhere, so it seems to be not needed anymore.